app/models/account.rb

```diff
@@ -438,98 +441,96 @@ def inboxes
     end
 
     def search_for(terms, limit = 10, offset = 0)
-      textsearch, query = generate_query_for_search(terms)
+      tsquery = generate_query_for_search(terms)
 
       sql = <<-SQL.squish
         SELECT
           accounts.*,
-          ts_rank_cd(#{textsearch}, #{query}, 32) AS rank
+          ts_rank_cd(#{TEXTSEARCH}, to_tsquery('simple', :tsquery), 32) AS rank
         FROM accounts
-        WHERE #{query} @@ #{textsearch}
+        WHERE to_tsquery('simple', :tsquery) @@ #{TEXTSEARCH}
           AND accounts.suspended_at IS NULL
           AND accounts.moved_to_account_id IS NULL
         ORDER BY rank DESC
-        LIMIT ? OFFSET ?
+        LIMIT :limit OFFSET :offset
       SQL
 
-      records = find_by_sql([sql, limit, offset])
+      records = find_by_sql([sql, limit: limit, offset: offset, tsquery: tsquery])
       ActiveRecord::Associations::Preloader.new.preload(records, :account_stat)
       records
     end
 
     def advanced_search_for(terms, account, limit = 10, following = false, offset = 0)
-      textsearch, query = generate_query_for_search(terms)
+      tsquery = generate_query_for_search(terms)
+      sql = advanced_search_for_sql_template(following)
+      records = find_by_sql([sql, id: account.id, limit: limit, offset: offset, tsquery: tsquery])
+      ActiveRecord::Associations::Preloader.new.preload(records, :account_stat)
+      records
+    end
+
+    def from_text(text)
+      return [] if text.blank?
 
+      text.scan(MENTION_RE).map { |match| match.first.split('@', 2) }.uniq.filter_map do |(username, domain)|
+        domain = begin
+          if TagManager.instance.local_domain?(domain)
+            nil
+          else
+            TagManager.instance.normalize_domain(domain)
+          end
+        end
+        EntityCache.instance.mention(username, domain)
+      end
+    end
+
+    private
+
+    def generate_query_for_search(unsanitized_terms)
+      terms = unsanitized_terms.gsub(DISALLOWED_TSQUERY_CHARACTERS, ' ')
+
+      # The final ":*" is for prefix search.
+      "'#{terms}':*"
```

Comment:
I think I mentioned this before: This does not yield the same results as `''' ' || #{terms} || ' ''' || ':*'`

Reply:
Which different results? `''' ' || #{terms} || ' ''' || ':*'` (when `#{terms}` get substituted by a string surrounded by single-quotes and not containing some characters) is a way to write the string `' #{terms} ':*` in SQL (it's tricky because of the single-quotes), `"' #{terms} ':*"` results in the same string, which cannot be pasted directly in SQL, but it's ok because it's used as a parameter to a parameterized query instead.

So the difference between the two strings is just the spaces surrounding `#{terms}`. I understand the query is from https://github.com/Casecommons/pg_search/blob/master/lib/pg_search/features/tsearch.rb#L100-L138 but I don't understand what those extra spaces would mean, I could not understand it from PostgreSQL's documentation and did not observe any difference on testing data.

Reply:
Since I could not understand the reason for the surrounding spaces in the built query, I ended up comparing the generated `tsqueries` (at the SQL level) using the following code:
```ruby
DISALLOWED_TSQUERY_CHARACTERS = /['?\\:‘’]/.freeze

def compare_tsqueries(unsanitized_terms)
  terms = unsanitized_terms.gsub(DISALLOWED_TSQUERY_CHARACTERS, ' ')
  inline_terms = Arel.sql(ActiveRecord::Base.connection.quote(terms))
  inline_tsquery = "to_tsquery('simple', ''' ' || #{inline_terms} || ' ''' || ':*')"
  tsquery_source = "'#{terms}':*"

  sql = "SELECT #{inline_tsquery} = to_tsquery('simple', $1)"
  ActiveRecord::Base.connection.select_all(sql, nil, [[nil, tsquery_source]]).rows[0][0]
end

def random_ascii_string(length)
  value = ''
  length.times do
    value  << (32 + rand(126-32)).chr
  end
  value
end

1000000.times.map { random_ascii_string(10) }.filter { |x| !compare_tsqueries(x) }
```

While I don't understand that result, it seems `to_tsquery('simple', query)` works slightly differently if the first term of query starts with `./`, `../` or `.. `. I don't know how to explain that change, so I'm introducing the spaces back, even though I do not think these discrepancies (one case searching for a leading `.` less) would cause any issue at all.